src/mutable_list.jl

```diff
@@ -0,0 +1,217 @@
+mutable struct ListNode{T}
+    data::T
+    prev::ListNode{T}
+    next::ListNode{T}
+    function ListNode{T}() where T
+        node = new{T}()
+        node.prev = node
+        node.next = node
+        return node
+    end
+    function ListNode{T}(data) where T
+        node = new{T}(data)
+        return node
+    end
+end
+
+mutable struct MutableLinkedList{T}
+    len::Int
+    front::ListNode{T}
+    back::ListNode{T}
+    function MutableLinkedList{T}() where T
+        l = new{T}()
+        l.len = 0
+        l.front = ListNode{T}()
+        l.back = ListNode{T}()
+        l.front.next = l.back
+        l.back.prev = l.front
+        return l
+    end
+end
+
+MutableLinkedList() = MutableLinkedList{Any}()
+
+function MutableLinkedList{T}(elts...) where T
+    l = MutableLinkedList{T}()
+    for elt in elts
+        push!(l, elt)
+    end
+    return l
+end
+
+iterate(l::MutableLinkedList) = l.len == 0 ? nothing : (l.front.next.data, l.front.next.next)
+iterate(l::MutableLinkedList, n::ListNode) = n.next == n ? nothing : (n.data, n.next)
+
+isempty(l::MutableLinkedList) = l.len == 0
+length(l::MutableLinkedList) = l.len
+collect(l::MutableLinkedList{T}) where T = T[x for x in l]
+eltype(l::MutableLinkedList{T}) where T = T
+first(l::MutableLinkedList) = l.front.next.data
+last(l::MutableLinkedList) = l.back.prev.data
+
+function ==(l1::MutableLinkedList, l2::MutableLinkedList)
+    typeof(l1) == typeof(l2) || return false
+    length(l1) == length(l2) || return false
+    for (i, j) in zip(l1, l2)
+        i == j || return false
+    end
+    return true
+end
+
+function map(f::Function, l::MutableLinkedList{T}) where T
+    l2 = MutableLinkedList{T}()
+    for h in l
+        push!(l2, f(h))
+    end
+    return l2
+end
+
+function filter(f::Function, l::MutableLinkedList{T}) where T
+    l2 = MutableLinkedList{T}()
+    for h in l
+        if f(h)
+            push!(l2, h)
+        end
+    end
+    return l2
+end
+
+function reverse(l::MutableLinkedList{T}) where T
+    l2 = MutableLinkedList{T}()
+    for h in l
+        pushfirst!(l2, h)
+    end
+    return l2
+end
+
+function copy(l::MutableLinkedList{T}) where T
+    l2 = MutableLinkedList{T}()
+    for h in l
+        push!(l2, h)
+    end
+    return l2
+end
+
+function getindex(l::MutableLinkedList, idx::Int)
+    0 < idx <= l.len || throw(ArgumentError("Index out of bounds"))
```

Comment:
Typically for bounds errors we should use Base's `@boundscheck` so that `@inbounds` knows how to skip bounds checking for `MutableLinkedList`s. So for example, I think this should be
```julia
@boundscheck 0 < idx <= l.len || throw(BoundsError(l, idx))
```

Reply:
(Same applies to other places you've implemented bounds checking)